File: cdd/LeetCode/LeetCode/inorderTraversal.cpp
#include <stdio.h>
#include <vector>
#include "TreeNode.cpp"
#include <stack>

using namespace std;

class Solution {
public:

	vector<int> res;
	

	vector<int> inorderTraversal(TreeNode* root){
		if (!root) {
			return res;
		}
		inorderTraversal(root->left);
		res.push_back(root->val);
		inorderTraversal(root->right);
		return res;
	}


	vector<int> inorderTraversal(TreeNode* root) {
		vector<int> res;
		stack<TreeNode*> stack;
		while (root != nullptr || !stack.empty()) {
			while (root != nullptr) {
				stack.push(root);
				root = root->left;
			}
			root = stack.top();
			stack.pop();
			res.push_back(root->val);
			root = root->right;
		}
		return res;
	}

};

File: python3/LeetCode/inorderTraversal.py
from typing import Optional, List

from LeetCode import TreeNode


class Solution:
    def inorderTraversal(self, root:TreeNode) -> List[int]:
        res = []
        def dfs(root):
            if not root:
                return
            dfs(root.left)
            res.append(root.val)
            dfs(root.right)
        dfs(root)
        return res

    def inorderTraversal(self, root:TreeNode) -> List[int]:
        res = []
        stack = []
        while root or stack:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            res.append(root.val)
            root = root.right
        return res